Add erc20 docs (#689)
OpenZeppelin/cairo-contracts: docs/modules/ROOT/pages/api/erc20.adoc

Sets `amount` as the allowance of `spender` over the caller's tokens.
Returns `true` on success.

Emits an <<Approval,Approval>> event.

This function panics if:

- The caller is the zero address.
- `spender` is the zero address.

[.contract-item]
[[IERC20-Transfer]]
==== `[.contract-item-name]#++Transfer++#++(from: ContractAddress, to: ContractAddress, value: u256)++` [.item-kind]#event#

Emitted when `value` tokens are moved from one address (`from`) to another (`to`).

Note that `value` may be zero.

[.contract-item]
[[IERC20-Approval]]
==== `[.contract-item-name]#++Approval++#++(owner: ContractAddress, spender: ContractAddress, value: u256)++` [.item-kind]#event#

Emitted when the allowance of a `spender` for an `owner` is set by a call to <<approve,approve>>.
`value` is the new allowance.

[.contract]
[[ERC20]]
=== `++ERC20++` link:https://github.com/OpenZeppelin/cairo-contracts/blob/cairo-2/src/token/erc20/erc20.cairo[{github-icon},role=heading-link]

[.hljs-theme-light.nopadding]
```javascript
use openzeppelin::token::erc20::ERC20;
```

Implementation of the <<IERC20,IERC20>> interface.

[.contract-index]
.Functions
--

> * xref:#constructor[`++constructor(name: felt252, symbol: felt252, initial_supply: u256, recipient: ContractAddress)++`]

Reviewer: As I said, I think it may be a good idea to add constructors, but I'm having second thoughts because they are not really part of the API (are not external methods, or internal), and they are not included in the Solidity docs version. IMO is better to design a way of presenting this that is not in the functions section, I'm leaning towards leaving this for a future iteration as well, wdty?

Author: >  I'm having second thoughts because they are not really part of the API (are not external methods, or internal)

We could set the `[.item-kind]` to `constructor` on the right side. If we like this, I think we should make it a point to always include it as the first function on the `Functions` list

> they are not included in the Solidity docs version

They are included though, no? https://docs.openzeppelin.com/contracts/4.x/api/token/erc20#ERC20-constructor-string-string-

> IMO is better to design a way of presenting this that is not in the functions section, I'm leaning towards leaving this for a future iteration as well,  wdty?

Hmm I'm actually leaning the other way. If we're insistent that the `constructor` doesn't fit in `Functions`, I think it's better to include it temporarily so users can still find information on it as opposed to completely omitting it. Wdyt?

Reviewer: I totally agree. Old message sorry, I got to the same point after a couple of iterations.

Author: Haha no worries, glad we're in agreement

:github-icon: pass:[<svg class="icon"><use href="#github-icon"/></svg>]
:eip20: https://eips.ethereum.org/EIPS/eip-20[EIP-20]
:erc20-guide: xref:erc20.adoc[ERC20 guide]

= ERC20

Reference of interfaces and utilities related to ERC20 contracts.

TIP: For an overview of ERC20, read our {erc20-guide}.

== Core

[.contract]
[[IERC20]]
=== `++IERC20++` link:https://github.com/OpenZeppelin/cairo-contracts/blob/cairo-2/src/token/erc20/interface.cairo#L6-L19[{github-icon},role=heading-link]

[.hljs-theme-light.nopadding]
```javascript
use openzeppelin::token::erc20::interface::IERC20;

> ```

Reviewer 2: do we really prefer the light theme here? cc/ @ericnordelo @andrew-fleming

Reviewer: Good question. Actually, I think I lean toward the dark version because makes it easier to note the separation between different sections. Not strong opinion though, I'm ok with light to keep consistency with the solidity doc site.

<img width="768" alt="Screenshot 2023-09-12 at 14 16 46" src="https://github.com/OpenZeppelin/cairo-contracts/assets/26070224/924ce4ec-85ba-488c-bec8-f8f0e2d0352f">

Author: +1 to the dark theme

Interface of the IERC20 standard as defined in {eip20}.

[.contract-index]
.Functions
--
* xref:#IERC20-total_supply[`++total_supply()++`]
* xref:#IERC20-balance_of[`++balance_of()++`]
* xref:#IERC20-allowance[`++allowance(owner: ContractAddress, spender: ContractAddress)++`]
* xref:#IERC20-transfer[`++transfer(recipient: ContractAddress, amount: u256)++`]
* xref:#IERC20-transfer_from[`++transfer_from(sender: ContractAddress, recipient: ContractAddress, amount: u256)++`]

> * xref:#IERC20-approve[`++approve(spender: ContractAddress, amount: u256)++`]

Reviewer 2: i thought we were skipping argument types in here ([like in Solidity](https://docs.openzeppelin.com/contracts/4.x/api/token/erc20#IERC20)). although i wouldn't skip return values

Reviewer 2: for details I agree, but to me return value is as important as the arguments

Author: Sounds good. No strong opinions on whether we should include the return values. I just followed the [Solidity docs](https://docs.openzeppelin.com/contracts/4.x/api/token/erc20#IERC20) approach. Happy to change it if we really think it's necessary

Reviewer 2: let's not block this PR and discuss it in #726 instead

Author: > for details I agree, but to me return value is as important as the arguments

Whoops, I missed this in my response^

> let's not block this PR and discuss it in https://github.com/OpenZeppelin/cairo-contracts/issues/726 instead

Sounds good

--

[.contract-index]
.Events
--
* xref:#IERC20-Transfer[`++Transfer(from: ContractAddress, to: ContractAddress, value: u256)++`]
* xref:#IERC20-Approval[`++Approval(owner: ContractAddress, spender: ContratAddress, value: u256)++`]
--

[.contract-item]
[[IERC20-total_supply]]
==== `[.contract-item-name]#++total_supply++#++() → u256++` [.item-kind]#external#

Returns the amount of tokens in existence.

[.contract-item]
[[IERC20-balance_of]]
==== `[.contract-item-name]#++balance_of++#++(account: ContractAddress) → u256++` [.item-kind]#external#

Returns the amount of tokens owned by `account`.

[.contract-item]
[[IERC20-allowance]]
==== `[.contract-item-name]#++allowance++#++(owner: ContractAddress, spender: ContractAddress) → u256++` [.item-kind]#external#

> Returns the remaining number of tokens that `spender` will be allowed to spend on behalf of `owner` through <<transfer_from,transfer_from>>. This is zero by default.

Reviewer 2: ```suggestion
Returns the remaining number of tokens that `spender` is allowed to spend on behalf of `owner` through <<transfer_from,transfer_from>>. This is zero by default.
```

This value changes when <<approve,approve>> or <<transfer_from,transfer_from>> are called.

[.contract-item]
[[IERC20-transfer]]
==== `[.contract-item-name]#++transfer++#++(recipient: ContractAddress, amount: u256) → bool++` [.item-kind]#external#

Moves `amount` tokens from the caller's token balance to `to`.

> Returns `true` on success.

Reviewer 2: ```suggestion
Returns `true` on success, reverts otherwise.
```

Emits a <<Transfer,Transfer>> event.

This function panics if:

- The caller is the zero address.
- `recipient` is the zero address.

[.contract-item]
[[IERC20-transfer_from]]
==== `[.contract-item-name]#++transfer_from++#++(sender: ContractAddress, recipient: ContractAddress, amount: u256) → bool++` [.item-kind]#external#

Moves `amount` tokens from `sender` to `recipient` using the allowance mechanism.
`amount` is then deducted from the caller's allowance.

> Returns `true` on success.

Reviewer 2: ```suggestion
Returns `true` on success, reverts otherwise.
```

Emits a <<Transfer,Transfer>> event.

This function panics if:

- `sender` is the zero address.
- `recipient` is the zero address.

[.contract-item]
[[IERC20-approve]]
==== `[.contract-item-name]#++approve++#++(spender: ContractAddress, amount: u256) → bool++` [.item-kind]#external#

Sets `amount` as the allowance of `spender` over the caller's tokens.

> Returns `true` on success.

Reviewer 2: ```suggestion
Returns `true` on success, reverts otherwise.
```

Emits an <<Approval,Approval>> event.

This function panics if:

- The caller is the zero address.
- `spender` is the zero address.

[.contract-item]
[[IERC20-Transfer]]
==== `[.contract-item-name]#++Transfer++#++(from: ContractAddress, to: ContractAddress, value: u256)++` [.item-kind]#event#

Emitted when `value` tokens are moved from one address (`from`) to another (`to`).

Note that `value` may be zero.

[.contract-item]
[[IERC20-Approval]]
==== `[.contract-item-name]#++Approval++#++(owner: ContractAddress, spender: ContractAddress, value: u256)++` [.item-kind]#event#

> Emitted when the allowance of a `spender` for an `owner` is set by a call to <<approve,approve>>.

Reviewer 2: ```suggestion
Emitted when the allowance of a `spender` for an `owner` is set.
```
usually we don't control when an Event is emitted. Users might use internal methods in many ways. For example this event can be emitted through `increase_allowance`, not just `approve`.

`value` is the new allowance.

[.contract]
[[ERC20]]
=== `++ERC20++` link:https://github.com/OpenZeppelin/cairo-contracts/blob/cairo-2/src/token/erc20/erc20.cairo[{github-icon},role=heading-link]

[.hljs-theme-light.nopadding]
```javascript
use openzeppelin::token::erc20::ERC20;
```

Implementation of the <<IERC20,IERC20>> interface.

> [.contract-index]

Reviewer 2: i think we should also include the camelCase functions as a subindex

* xref:#balance_of[`++balance_of(self, account)++`]
* xref:#allowance[`++allowance(self, owner, spender)++`]
* xref:#transfer[`++transfer(self, recipient, amount)++`]
* xref:#transfer_from[`++transfer_from(self, sender, recipient, amount)++`]
* xref:#approve[`++approve(self, spender, amount)++`]

[.contract-subindex-inherited]
.InternalImpl

* xref:#initializer[`++initializer(self, name_, symbol_)++`]
* xref:#_increase_allowance[`++_increase_allowance(self, spender, added_value)++`]
* xref:#_decrease_allowance[`++_decrease_allowance(self, spender, subtracted_value)++`]
* xref:#_mint[`++_mint(self, recipient, amount)++`]
* xref:#_burn[`++_burn(self, account, amount)++`]
* xref:#_approve[`++_approve(self, owner, spender, amount)++`]
* xref:#_transfer[`++_transfer(self, sender, recipient, amount)++`]
* xref:#_spend_allowance[`++_spend_allowance(self, owner, spender, amount)++`]

--

[.contract-index]
.Events
--
* xref:#Transfer[`++Transfer(from: ContractAddress, to: ContractAddress, value: u256)++`]
* xref:#Approval[`++Approval(owner: ContractAddress, spender: ContratAddress, value: u256)++`]
--

[#ERC20-Functions]
==== Functions

[.contract-item]
[[constructor]]

> ==== `[.contract-item-name]#++constructor++#++(ref self: ContractState, name_: felt252, symbol: felt252, initial_supply: u256, recipient: ContractAddress)++` [.item-kind]#constructor#

Reviewer 2: ```suggestion
==== `[.contract-item-name]#++constructor++#++(ref self: ContractState, name: felt252, symbol: felt252, initial_supply: u256, recipient: ContractAddress)++` [.item-kind]#constructor#
```
i think the constructor uses regular `name`?

Sets both the token name and symbol and mints `initial_supply` to `recipient`.
Note that the token name and symbol are immutable once set through the constructor.

[.contract-item]
[[name]]
==== `[.contract-item-name]#++name++#++(@self: ContractState) → felt252++` [.item-kind]#external#

Returns the name of the token.

[.contract-item]
[[symbol]]
==== `[.contract-item-name]#++symbol++#++(@self: ContractState) → felt252++` [.item-kind]#external#

Returns the ticker symbol of the token.

[.contract-item]
[[decimals]]
==== `[.contract-item-name]#++decimals++#++(@self: ContractState) → u8++` [.item-kind]#external#

> Returns the number of decimals the token uses - e.g. 8 means to divide the token amount by 100000000 to get its user representation.

Reviewer 2: ```suggestion
Returns the number of decimals the token uses - e.g. `8` means to divide the token amount by `100000000` to get its user-readable representation.
```

For example, if `decimals` equals `2`, a balance of `505` tokens should
be displayed to a user as `5.05` (`505 / 10 ** 2`).

> Tokens usually opt for a value of 18, imitating the relationship between

Reviewer 2: ```suggestion
Tokens usually opt for a value of `18`, imitating the relationship between
```

See <<IERC20-balance_of,IERC20::balance_of>>.

[.contract-item]
[[allowance]]
==== `[.contract-item-name]#++allowance++#++(@self: ContractState, owner: ContractAddress, spender: ContractAddress) → u256++` [.item-kind]#external#

See <<IERC20-allowance,IERC20::allowance>>.

[.contract-item]
[[transfer]]
==== `[.contract-item-name]#++transfer++#++(ref self: ContractState, recipient: ContractAddress, amount: u256) → bool++` [.item-kind]#external#

See <<IERC20-transfer,IERC20::transfer>>.

[.contract-item]
[[transfer_from]]
==== `[.contract-item-name]#++transfer_from++#++(ref self: ContractState, sender: ContractAddress, recipient: ContractAddress, amount: u256) → bool++` [.item-kind]#external#

See <<IERC20-transfer_from,IERC20::transfer_from>>.

[.contract-item]
[[approve]]
==== `[.contract-item-name]#++approve++#++(ref self: ContractState, spender: ContractAddress, amount: u256) → bool++` [.item-kind]#external#

See <<IERC20-approve,IERC20::approve>>.

[.contract-item]
[[increase_allowance]]
==== `[.contract-item-name]#++increase_allowance++#++(ref self: ContractState, spender: ContractAddress, added_value: u256) → bool++` [.item-kind]#external#

> Atomically increases the allowance granted to `spender` by the caller.

Reviewer 2: ```suggestion
Increases the allowance granted from the caller to `spender` by `added_value`
```

> Returns `true` on success.

Reviewer 2: ```suggestion
Returns `true` on success, reverts otherwise.
```

Emits an <<Approval,Approval>> event.

This function panics if:

- The caller is the zero address.
- `spender` is the zero address.

[.contract-item]
[[decrease_allowance]]
==== `[.contract-item-name]#++decrease_allowance++#++(ref self: ContractState, spender: ContractAddress, subtracted_value: u256) → bool++` [.item-kind]#external#

> Atomically decreases the allowance granted to `spender` by the caller.

Reviewer 2: ```suggestion
Decreases the allowance granted from the caller to `spender` by `subtracted_value`
```

Returns `true` on success.

Emits an <<Approval,Approval>> event.

This function panics if:

- The caller is the zero address.
- `spender` is the zero address.

> 

Reviewer 2: should we add some title or separator between external and internal functions? (and any other set of functions, such as internal, external, maybe constructor/initializers?

Author: I think that's a good idea. I added new headers here: https://github.com/OpenZeppelin/cairo-contracts/pull/689/commits/108dfbffa475461c1b6343a081d7b8e1b407c781

Let me know what you think

Reviewer: I think is better is we separate the header section, no the ones with the function description:

This looks redundant to me, the external is already in the top-right corner.

<img width="769" alt="Screenshot 2023-09-13 at 15 24 01" src="https://github.com/OpenZeppelin/cairo-contracts/assets/26070224/ce6ff4a2-704b-4d92-a0c6-abc0e14cc70b">

I think something like this would be better, keeping the descriptions as it is.

<img width="752" alt="Screenshot 2023-09-13 at 15 25 15" src="https://github.com/OpenZeppelin/cairo-contracts/assets/26070224/624ee2bf-5e0f-4122-ab8c-bc4e328ce5b4">

Reviewer 2: i like how it looks now, but i also don't think there's any problem with a bit of added redundancy. docsites are reference sites, not books you read in order, and redundancy is often a good idea to maximize the value obtained by reading a small autocontained version.

i think this redundancy is especially useful for navigation, which is usually hard in API docs

Author: I don't disagree with Marto's perspective especially for larger contracts a la ERC721. I say let's add them in. We can always change it later

Reviewer: Got it, I will update accordingly, maybe in the upgrades PR, to do it in one place if you agree (that PR contains the rest of my doc PRs except overview)

Author: Good idea, and merge that after the others. Works for me!

Reviewer 2: I really like how it looks now!

I think it still looks a bit weird that `increase_allowance` and `decrease_allowance` are above the rest, and with a different color. should we add them to their own impl and place them below? i'd prioritize `IERC20` ones

![image](https://github.com/OpenZeppelin/cairo-contracts/assets/1703831/a32c7a89-283e-4c84-9fac-1b400fcf6362)

[.contract-item]
[[initializer]]
==== `[.contract-item-name]#++initializer++#++(ref self: ContractState, name_: felt252, symbol_: felt252)++` [.item-kind]#internal#

Initializes the contract by setting the token name and symbol.
This should be used inside of the contract's constructor.

[.contract-item]
[[_increase_allowance]]
==== `[.contract-item-name]#++_increase_allowance++#++(ref self: ContractState, spender: ContractAddress, added_value: u256)++` [.item-kind]#internal#

> Atomically increases the allowance granted to `spender` by the caller.

Reviewer 2: ```suggestion
Increases the allowance granted from the caller to `spender` by `added_value`
```

Emits an <<Approval,Approval>> event.

[.contract-item]
[[_decrease_allowance]]
==== `[.contract-item-name]#++_decrease_allowance++#++(ref self: ContractState, spender: ContractAddress, subtracted_value: u256)++` [.item-kind]#internal#

> Atomically decreases the allowance granted to `spender` by the caller.

Reviewer 2: ```suggestion
Decreases the allowance granted from the caller to `spender` by `subtracted_value`
```

Emits an <<Approval,Approval>> event.

[.contract-item]
[[_mint]]
==== `[.contract-item-name]#++_mint++#++(ref self: ContractState, recipient: ContractAddress, amount: u256)++` [.item-kind]#internal#

> Creates an `amount` number of tokens and assigns them to `recipient`, by transferring them from the zero address.

Reviewer 2: ```suggestion
Creates an `amount` number of tokens and assigns them to `recipient`.
```

Emits a <<Transfer,Transfer>> event with `from` being the zero address.

This function panics if:

- `recipient` is the zero address.

[.contract-item]
[[_burn]]
==== `[.contract-item-name]#++_burn++#++(ref self: ContractState, account: ContractAddress, amount: u256)++` [.item-kind]#internal#

> Destroys `amount` number of tokens from `account`, by transferring them to the zero address.

Reviewer 2: ```suggestion
Destroys `amount` number of tokens from `account`.
```

Emits a <<Transfer,Transfer>> event with `to` set to the zero address.

This function panics if:

- `account` is the zero address.

[.contract-item]
[[_approve]]
==== `[.contract-item-name]#++_approve++#++(ref self: ContractState, owner: ContractAddress, spender: ContractAddress, amount: u256)++` [.item-kind]#internal#

> Sets `amount` as the allowance of `spender` over the `owner` tokens.

Reviewer 2: ```suggestion
Sets `amount` as the allowance of `spender` over `owner`'s tokens.
```

> This internal function is equivalent to `approve`, and can be used to e.g. set automatic allowances for certain subsystems, etc.

Reviewer 2: can you clarify this example?

Emits an <<Approval,Approval>> event.

This function panics if:

- `owner` is the zero address.
- `spender` is the zero address.

[.contract-item]
[[_transfer]]
==== `[.contract-item-name]#++_transfer++#++(ref self: ContractState, sender: ContractAddress, recipient: ContractAddress, amount: u256)++` [.item-kind]#internal#

Sets `amount` as the allowance of `spender` over the `owner` tokens.

> This internal function is equivalent to `approve`, and can be used to e.g. set automatic allowances for certain subsystems, etc.

Reviewer 2: this is `_approve`'s section

:github-icon: pass:[<svg class="icon"><use href="#github-icon"/></svg>]
:eip20: https://eips.ethereum.org/EIPS/eip-20[EIP-20]
:erc20-guide: xref:erc20.adoc[ERC20 guide]
:casing-discussion: https://github.com/OpenZeppelin/cairo-contracts/discussions/34[here]

= ERC20

Reference of interfaces and utilities related to ERC20 contracts.

TIP: For an overview of ERC20, read our {erc20-guide}.

== Core

[.contract]
[[IERC20]]
=== `++IERC20++` link:https://github.com/OpenZeppelin/cairo-contracts/blob/cairo-2/src/token/erc20/interface.cairo#L6-L19[{github-icon},role=heading-link]

[.hljs-theme-dark]
```javascript
use openzeppelin::token::erc20::interface::IERC20;
```

Interface of the IERC20 standard as defined in {eip20}.

[.contract-index]
.Functions
--

> * xref:#IERC20::name[`++name()++`]
> * xref:#IERC20::symbol[`++symbol()++`]
> * xref:#IERC20::decimals[`++decimals()++`]
> * xref:#IERC20::total_supply[`++total_supply()++`]
> * xref:#IERC20::balance_of[`++balance_of()++`]
> * xref:#IERC20::allowance[`++allowance(owner, spender)++`]
> * xref:#IERC20::transfer[`++transfer(recipient, amount)++`]
> * xref:#IERC20::transfer_from[`++transfer_from(sender, recipient, amount)++`]
> * xref:#IERC20::approve[`++approve(spender, amount)++`]

Reviewer: ```suggestion
* xref:#IERC20-name[`++name()++`]
* xref:#IERC20-symbol[`++symbol()++`]
* xref:#IERC20-decimals[`++decimals()++`]
* xref:#IERC20-total_supply[`++total_supply()++`]
* xref:#IERC20-balance_of[`++balance_of()++`]
* xref:#IERC20-allowance[`++allowance(owner, spender)++`]
* xref:#IERC20-transfer[`++transfer(recipient, amount)++`]
* xref:#IERC20-transfer_from[`++transfer_from(sender, recipient, amount)++`]
* xref:#IERC20-approve[`++approve(spender, amount)++`]
```

This is the pattern I've been using (the one from Solidity doc site), and I think is more URL friendly than the colons. This ids are shown in the URL.

Emits an <<Approval,Approval>> event.

Requirements:

- `spender` cannot be the zero address.

[.contract-item]
[[IERC20::Transfer]]
==== `[.contract-item-name]#++Transfer++#++(from: ContractAddress, to: ContractAddress, value: u256)++` [.item-kind]#event#

Emitted when `value` tokens are moved from one address (`from`) to another (`to`).

Note that `value` may be zero.

[.contract-item]
[[IERC20::Approval]]
==== `[.contract-item-name]#++Approval++#++(owner: ContractAddress, spender: ContractAddress, value: u256)++` [.item-kind]#event#

Emitted when the allowance of a `spender` for an `owner` is set.
`value` is the new allowance.

[.contract]
[[ERC20]]
=== `++ERC20++` link:https://github.com/OpenZeppelin/cairo-contracts/blob/cairo-2/src/token/erc20/erc20.cairo[{github-icon},role=heading-link]

[.hljs-theme-dark]
```javascript
use openzeppelin::token::erc20::ERC20;
```

Implementation of the <<IERC20,IERC20>> interface.

[.contract-index]
.External Functions
--
* xref:#increase_allowance[`++increase_allowance(self, spender, added_value)++`]
* xref:#decrease_allowance[`++decrease_allowance(self, spender, subtracted_value)++`]

[.contract-subindex-inherited]
.IERC20

> * xref:#name[`++name(self)++`]
> * xref:#symbol[`++symbol(self)++`]
> * xref:#decimals[`++decimals(self)++`]

Reviewer: This three methods are missing below.

Also, we should always scope the Ids, I think we should always scope the IDs, in case we are more modules.

```suggestion
* xref:#ERC20-name[`++name(self)++`]
* xref:#ERC20-symbol[`++symbol(self)++`]
* xref:#ERC20-decimals[`++decimals(self)++`]
```

= ERC20

Reference of interfaces and utilities related to ERC20 contracts.

TIP: For an overview of ERC20, read our {erc20-guide}.

== Core

[.contract]
[[IERC20]]
=== `++IERC20++` link:https://github.com/OpenZeppelin/cairo-contracts/blob/cairo-2/src/token/erc20/interface.cairo#L6-L19[{github-icon},role=heading-link]

[.hljs-theme-dark]
```javascript
use openzeppelin::token::erc20::interface::IERC20;
```

Interface of the IERC20 standard as defined in {eip20}.

[.contract-index]
.Functions
--
* xref:#IERC20-name[`++name()++`]
* xref:#IERC20-symbol[`++symbol()++`]
* xref:#IERC20-decimals[`++decimals()++`]
* xref:#IERC20-total_supply[`++total_supply()++`]
* xref:#IERC20-balance_of[`++balance_of()++`]
* xref:#IERC20-allowance[`++allowance(owner, spender)++`]
* xref:#IERC20-transfer[`++transfer(recipient, amount)++`]
* xref:#IERC20-transfer_from[`++transfer_from(sender, recipient, amount)++`]
* xref:#IERC20-approve[`++approve(spender, amount)++`]
--

[.contract-index]
.Events
--
* xref:#IERC20-Transfer[`++Transfer(from, to, value)++`]
* xref:#IERC20-Approval[`++Approval(owner, spender, value)++`]
--

> 

Reviewer: ```suggestion

[#IERC20-Functions]
==== Functions

```

Emits a <<ERC20-Transfer,Transfer>> event.

Requirements:

- `recipient` cannot be the zero address.
- The caller must have a balance of at least `amount`.

[.contract-item]
[[IERC20-transfer_from]]
==== `[.contract-item-name]#++transfer_from++#++(sender: ContractAddress, recipient: ContractAddress, amount: u256) → bool++` [.item-kind]#external#

Moves `amount` tokens from `sender` to `recipient` using the allowance mechanism.
`amount` is then deducted from the caller's allowance.
Returns `true` on success, reverts otherwise.

Emits a <<ERC20-Transfer,Transfer>> event.

Requirements:

- `sender` cannot be the zero address.
- `recipient` cannot be the zero address.
- `sender` must have a balance of at least `amount`.
- The caller must have allowance for ``sender``'s tokens of at least `amount`.

[.contract-item]
[[IERC20-approve]]
==== `[.contract-item-name]#++approve++#++(spender: ContractAddress, amount: u256) → bool++` [.item-kind]#external#

Sets `amount` as the allowance of `spender` over the caller's tokens.
Returns `true` on success, reverts otherwise.

Emits an <<ERC20-Approval,Approval>> event.

Requirements:

- `spender` cannot be the zero address.

> 

Reviewer: ```suggestion

[#IERC20-Events]
==== Events

```

[.contract-subindex-inherited]
.ERC20Camel

* xref:#ERC20-totalSupply[`++totalSupply(self)++`]
* xref:#ERC20-balanceOf[`++balanceOf(self, account)++`]
* xref:#ERC20-transferFrom[`++transferFrom(self, sender, recipient, amount)++`]
* xref:#ERC20-increaseAllowance[`++increaseAllowance(self, spender, addedValue)++`]
* xref:#ERC20-decreaseAllowance[`++decreaseAllowance(self, spender, subtractedValue)++`]
--

[.contract-index]
.Internal functions
--

[.contract-subindex-inherited]
.InternalImpl

* xref:#ERC20-initializer[`++initializer(self, name, symbol)++`]
* xref:#ERC20-_increase_allowance[`++_increase_allowance(self, spender, added_value)++`]
* xref:#ERC20-_decrease_allowance[`++_decrease_allowance(self, spender, subtracted_value)++`]
* xref:#ERC20-_mint[`++_mint(self, recipient, amount)++`]
* xref:#ERC20-_burn[`++_burn(self, account, amount)++`]
* xref:#ERC20-_approve[`++_approve(self, owner, spender, amount)++`]
* xref:#ERC20-_transfer[`++_transfer(self, sender, recipient, amount)++`]
* xref:#ERC20-_spend_allowance[`++_spend_allowance(self, owner, spender, amount)++`]

--

[.contract-index]
.Events
--

> * xref:#ERC20-Transfer[`++Transfer(from: ContractAddress, to: ContractAddress, value: u256)++`]
> * xref:#ERC20-Approval[`++Approval(owner: ContractAddress, spender: ContratAddress, value: u256)++`]

Reviewer: ```suggestion
* xref:#ERC20-Transfer[`++Transfer(from, to, value)++`]
* xref:#ERC20-Approval[`++Approval(owner, spender, value)++`]
```

See <<IERC20-transfer,IERC20::transfer>>.

Requirements:

- `recipient` cannot be the zero address.
- The caller must have a balance of at least `amount`.

[.contract-item]
[[ERC20-transfer_from]]
==== `[.contract-item-name]#++transfer_from++#++(ref self: ContractState, sender: ContractAddress, recipient: ContractAddress, amount: u256) → bool++` [.item-kind]#external#

See <<IERC20-transfer_from,IERC20::transfer_from>>.

Requirements:

> - `sender` cannot be the zero address.
> - `recipient` cannot be the zero address.
> - `sender` must have a balance of at least `amount`.
> - The caller must have allowance for ``sender``'s tokens of at least `amount`.
> 

Reviewer 2: ```suggestion
- `sender` cannot be the zero address.
- `sender` must have a balance of at least `amount`.
- `recipient` cannot be the zero address.
- The caller must have allowance for ``sender``'s tokens of at least `amount`.

```
what do you think about this? it groups together all reqs for `sender`

Author: Yeah, it's visually better. +1

==== `[.contract-item-name]#++_approve++#++(ref self: ContractState, owner: ContractAddress, spender: ContractAddress, amount: u256)++` [.item-kind]#internal#

Sets `amount` as the allowance of `spender` over ``owner``'s tokens.

> This internal function does not include permissions but can be useful for instances like setting automatic allowances on behalf of other addresses.

Reviewer 2: ```suggestion
This internal function does not check for access permissions but can be useful as a building block, for example to implement automatic allowances on behalf of other addresses.
```

[[ERC20-_transfer]]
==== `[.contract-item-name]#++_transfer++#++(ref self: ContractState, sender: ContractAddress, recipient: ContractAddress, amount: u256)++` [.item-kind]#internal#

Moves `amount` of tokens from `from` to `to`.

> This internal function does not include permissions but can be useful for instances like implementing automatic token fees, slashing mechanisms, etc.

Reviewer 2: ```suggestion
This internal function does not check for access permissions but can be useful as a building block, for example to implement automatic token fees, slashing mechanisms, etc.
```
